### DISCUSSION TITLE
4.0.0-RC11 - Blizzard

### DISCUSSION THREAD

**Discussion by byRoadrunner** [CloudNetService/CloudNet-v3#1532] [Releases]
# Update 4.0.0-RC11 - Blizzard #

We are pleased to announce the eleventh release candidate for CloudNet 4.0. This release includes fixes for reported issues and some improvements. We urge all users to install the update, as we will no longer provide support for RC10. Users who want to switch from 3.4.X to 4.0 can find instructions in the [release information for RC1](https://github.com/CloudNetService/CloudNet-v3/releases/tag/4.0.0-RC1).

Please also see the changelog for [RC6](https://github.com/CloudNetService/CloudNet-v3/releases/tag/4.0.0-RC6), this contains the detailed information about the switch to Dependency Injection in the API.

Cheers!
(**Please remember, CloudNet is provided as-is - we are not responsible for data loss or corruption. You are encouraged to back up your files before any updates!**)

# Upgrading
Due to some changes the upgrading step is a bit more complicated this time. 
* CloudNet requires JDK 23 everywhere now.
* Velocity 3.3.0-SNAPSHOT is required now.

After installing JDK 23 download the [CloudNet.zip](https://github.com/CloudNetService/CloudNet/releases/download/4.0.0-RC11/CloudNet.zip) and replace your launcher.jar with the one from the archive.

As announced the CloudPerms module and thus the SimpleNameTags and Chat plugins are not available anymore. Please delete the CloudPerms module before starting.

# Changelog

## Breaking API
* The `Task` class was replaced with the CompletableFutures from the standard library.  Some helper methods can be found in the `TaskUtil` class.

## Improvements
* Minecraft 1.21.1 is now fully supported (#1477)
* Limbo Minecraft servers are now supported (#1412)
* ServiceRemoteInclusions have a new cache strategy option to determine how to cache them (#1413)
* NPCs can now be rotated using the new `rotate` command (#1497)
* The networking was fully rewritten to the new JDK 22 classfile api (#1433)
* The terminal is now using a new implementation (#1441)
* WorldEdit and FAWE can now be used in older Minecraft Versions together with JDK 22 (#1429)
* New cloudnet luckperms plugin to bring tasks and groups into luckperms (#1258)

## Fixes
* The npc command sets the showfullservices option now correctly.
* Fabric fake players are not detected as real players anymore.
* Skipping the proxy fallback question is now working again.

<details>
<summary>Dependencies</summary>

| artifact id | usage                                                                                                        |
|-------------|--------------------------------------------------------------------------------------------------------------|
| driver      | When developing plugins or modules, available on all platforms.                                              |
| node        | When developing a module which needs more specific access to node functions than the driver can offer.       |
| wrapper-jvm | When developing plugins which need some more specific access to the service api than the driver can offer.   |
| bridge      | When trying to catch player events and/or interacting with players regardless where they are on the network. |
| bom         | When you want to import all dependencies with the same version based on the imported bill of materials.      |

To add the CloudNet dependency using gradle:

```groovy
repositories {
  // ensure maven central is added
  mavenCentral()
}

dependencies {
  // optional - you can also specify versions directly
  implementation platform('eu.cloudnetservice.cloudnet:bom:4.0.0-RC11')
  compileOnly 'eu.cloudnetservice.cloudnet:driver'

  // without bom
  compileOnly 'eu.cloudnetservice.cloudnet:driver:4.0.0-RC11'
}
```

To add the CloudNet dependency using maven:

```xml
<!-- optional - you can also specify versions directly -->
<dependencyManagement>
  <dependencies>
    <dependency>
      <groupId>eu.cloudnetservice.cloudnet</groupId>
      <artifactId>bom</artifactId>
      <version>4.0.0-RC11</version>
      <type>pom</type>
      <scope>import</scope>
    </dependency>
  </dependencies>
</dependencyManagement>

<dependencies>
  <dependency>
    <groupId>eu.cloudnetservice.cloudnet</groupId>
    <artifactId>driver</artifactId>
    <version>4.0.0-RC11</version> <!-- only needed when bom is not used -->
    <scope>provided</scope>
  </dependency>
</dependencies>
```
</details>

## 🔗 Links
* Support Discord: https://discord.cloudnetservice.eu/
* Javadocs: https://cloudnetservice.eu/cloudnet/docs/4.0.0-RC11/
* Maven Dependencies: https://central.sonatype.com/search?q=eu.cloudnetservice.cloudnet
* If you found any issue, please report it on our issue tracker: https://github.com/CloudNetService/CloudNet/issues

<hr /><em>This discussion was created from the release <a href='https://github.com/CloudNetService/CloudNet/releases/tag/4.0.0-RC11'>4.0.0-RC11 - Blizzard</a>.</em>

**Comment by GiantTreeLP** [comment]
I guess there is a typo in the release notes: 
```diff
--- Limbo Minecraft servers are not supported 
+++ Limbo Minecraft servers are now supported
```

**Reply by derklaro** [reply]
😄 fixed
